### PR TITLE
[new release] dns-lwt, dns, dns-async, dns-lwt-unix and mirage-dns (1.1.3)

### PR DESCRIPTION
**Pull request by avsm**
DNS implementation in portable Lwt

- Project page: <a href="https://github.com/mirage/ocaml-dns">https://github.com/mirage/ocaml-dns</a>
- Documentation: <a href="https://mirage.github.io/ocaml-dns/">https://mirage.github.io/ocaml-dns/</a>

##### CHANGES:

* Support domain-name.0.3.0 interface, which bumps the minimum
  OCaml version supported to 4.04 due to that dependency (@avsm)
* Fix tests with recent OCaml (use mmap/bigarray-compat) (@avsm)

### CODE REVIEW

File: packages/dns-async/dns-async.1.1.3/opam [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+and server resolver using the Jane Street Async library.
+"""
+license: "ISC"
+tags: ["org:mirage" "org:xapi-project"]
+homepage: "https://github.com/mirage/ocaml-dns"
+doc: "https://mirage.github.io/ocaml-dns/"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {build}
```
Comment (kit-ty-kate) [inline packages/dns-async/dns-async.1.1.3/opam]:
```suggestion
  "dune"
```

File: packages/dns-lwt-unix/dns-lwt-unix.1.1.3/opam [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+(as supported by Lwt_unix upstream).
+"""
+license: "ISC"
+tags: ["org:mirage" "org:xapi-project"]
+homepage: "https://github.com/mirage/ocaml-dns"
+doc: "https://mirage.github.io/ocaml-dns/"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {build}
```
Comment (kit-ty-kate) [inline packages/dns-lwt-unix/dns-lwt-unix.1.1.3/opam]:
```suggestion
  "dune"
```

File: packages/dns-lwt/dns-lwt.1.1.3/opam [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+opam package, and in theory you could compile this one to JavaScript
+as well using js_of_ocaml.
+"""
+tags: ["org:mirage" "org:xapi-project"]
+homepage: "https://github.com/mirage/ocaml-dns"
+doc: "https://mirage.github.io/ocaml-dns/"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune" {build}
```
Comment (kit-ty-kate) [inline packages/dns-lwt/dns-lwt.1.1.3/opam]:
```suggestion
  "dune"
```

File: packages/mirage-dns/mirage-dns.1.1.3/opam [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+homepage: "https://github.com/mirage/ocaml-dns"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+synopsis: "DNS implementation for the MirageOS unikernel framework"
+description: """
+This is an implementation of a DNS server and client resolver
+for the [MirageOS unikernel framework](https://mirage.io).
+"""
+depends: [
+  "ocaml" {>= "4.05.0"}
+  "dune" {build}
```
Comment (kit-ty-kate) [inline packages/mirage-dns/mirage-dns.1.1.3/opam]:
```suggestion
  "dune"
```

File: packages/dns/dns.1.1.3/opam [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+- dns-async using the Jane Street Async library
+- mirage-dns for the MirageOS unikernel framework
+"""
+tags: ["org:mirage" "org:xapi-project"]
+homepage: "https://github.com/mirage/ocaml-dns"
+doc: "https://mirage.github.io/ocaml-dns/"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune"
```
Comment (kit-ty-kate) [inline packages/dns/dns.1.1.3/opam]:
```suggestion
  "dune" {>= "1.2"}
```
According to the `dune-package` file, the version of dune required is 1.2

**Comment by kit-ty-kate** [comment]
Tests fail on OCaml 4.03:
```
#=== ERROR while compiling dns-lwt-unix.1.1.3 =================================#
# context              2.0.4 | linux/x86_64 | ocaml-base-compiler.4.03.0 | pinned(https://github.com/mirage/ocaml-dns/releases/download/v1.1.3/dns-v1.1.3.tbz)
# path                 ~/.opam/4.03/.opam-switch/build/dns-lwt-unix.1.1.3
# command              ~/.opam/4.03/bin/dune runtest -p dns-lwt-unix -j 2
# exit-code            1
# env-file             ~/.opam/log/dns-lwt-unix-15-3ec997.env
# output-file          ~/.opam/log/dns-lwt-unix-15-3ec997.out
### output ###
#     ocamlopt lib_test/unix/time_server.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.03/bin/ocamlopt.opt -w -40 -g -o lib_test/unix/time_server.exe -I /home/opam/.opam/4.03/lib/base64 -I /home/opam/.opam/4.03/lib/bytes -I /home/opam/.opam/4.03/lib/cstruct -I /home/opam/.opam/4.03/lib/dns -I /home/opam/.opam/4.03/lib/dns-lwt -I /home/opam/.opam/4.03/lib/hashcons -I /home/opam/.opam/4.03/lib/ipaddr -I /home/opam/.opam/4.03/lib/lwt -I /home/opam/.opam/4.03/lib/mirage-profile -I /home/opam/.opam/4.03/lib/ocplib-endian -I /home/opam/.opam/4.03/lib/re -I /home/opam/.opam/4.03/lib/re/posix -I /home/opam/.opam/4.03/lib/re/str -I /home/opam/.opam/4.03/lib/result -I /home/opam/.opam/4.03/lib/seq -I /home/opam/.opam/4.03/lib/sexplib -I /home/opam/.opam/4.03/lib/stringext -I /home/opam/.opam/4.03/lib/uri -I lwt-unix /home/opam/.opam/4.03/lib/ocaml/unix.cmxa /home/opam/.opam/4.03/lib/ocaml/bigarray.cmxa /home/opam/.opam/4.03/lib/sexplib/sexplib.cmxa /home/opam/.opam/4.03/lib/cstruct/cstruct.cmxa /home/opam/.opam/4.03/lib/seq/seq.cmxa /home/opam/.opam/4.03/lib/re/re.cmxa /home/opam/.opam/4.03/lib/re/str/re_str.cmxa /home/opam/.opam/4.03/lib/ipaddr/ipaddr.cmxa /home/opam/.opam/4.03/lib/re/posix/re_posix.cmxa /home/opam/.opam/4.03/lib/stringext/stringext.cmxa /home/opam/.opam/4.03/lib/uri/uri.cmxa /home/opam/.opam/4.03/lib/base64/base64.cmxa /home/opam/.opam/4.03/lib/uri/services.cmxa /home/opam/.opam/4.03/lib/hashcons/hashcons.cmx /home/opam/.opam/4.03/lib/result/result.cmxa /home/opam/.opam/4.03/lib/dns/dns.cmxa /home/opam/.opam/4.03/lib/lwt/lwt.cmxa /home/opam/.opam/4.03/lib/ocplib-endian/ocplib_endian.cmxa /home/opam/.opam/4.03/lib/ocplib-endian/bigstring.cmxa /home/opam/.opam/4.03/lib/mirage-profile/mirage_profile.cmxa /home/opam/.opam/4.03/lib/dns-lwt/dns_lwt.cmxa /home/opam/.opam/4.03/lib/ipaddr/ipaddr_unix.cmxa /home/opam/.opam/4.03/lib/lwt/lwt-log.cmxa /home/opam/.opam/4.03/lib/lwt/lwt-unix.cmxa lwt-unix/dns_lwt_unix.cmxa lib_test/unix/.time_server.eobjs/native/time_server.cmx)
# File "_none_", line 1:
# Error: Files lwt-unix/dns_lwt_unix.cmxa
#        and /home/opam/.opam/4.03/lib/dns-lwt/dns_lwt.cmxa
#        make inconsistent assumptions over implementation Dns_server
#     ocamlopt examples/server.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.03/bin/ocamlopt.opt -w -40 -g -o examples/server.exe -I /home/opam/.opam/4.03/lib/base64 -I /home/opam/.opam/4.03/lib/bytes -I /home/opam/.opam/4.03/lib/cstruct -I /home/opam/.opam/4.03/lib/dns -I /home/opam/.opam/4.03/lib/dns-lwt -I /home/opam/.opam/4.03/lib/hashcons -I /home/opam/.opam/4.03/lib/ipaddr -I /home/opam/.opam/4.03/lib/lwt -I /home/opam/.opam/4.03/lib/mirage-profile -I /home/opam/.opam/4.03/lib/ocplib-endian -I /home/opam/.opam/4.03/lib/re -I /home/opam/.opam/4.03/lib/re/posix -I /home/opam/.opam/4.03/lib/re/str -I /home/opam/.opam/4.03/lib/result -I /home/opam/.opam/4.03/lib/seq -I /home/opam/.opam/4.03/lib/sexplib -I /home/opam/.opam/4.03/lib/stringext -I /home/opam/.opam/4.03/lib/uri -I lwt-unix /home/opam/.opam/4.03/lib/ocaml/unix.cmxa /home/opam/.opam/4.03/lib/ocaml/bigarray.cmxa /home/opam/.opam/4.03/lib/sexplib/sexplib.cmxa /home/opam/.opam/4.03/lib/cstruct/cstruct.cmxa /home/opam/.opam/4.03/lib/seq/seq.cmxa /home/opam/.opam/4.03/lib/re/re.cmxa /home/opam/.opam/4.03/lib/re/str/re_str.cmxa /home/opam/.opam/4.03/lib/ipaddr/ipaddr.cmxa /home/opam/.opam/4.03/lib/re/posix/re_posix.cmxa /home/opam/.opam/4.03/lib/stringext/stringext.cmxa /home/opam/.opam/4.03/lib/uri/uri.cmxa /home/opam/.opam/4.03/lib/base64/base64.cmxa /home/opam/.opam/4.03/lib/uri/services.cmxa /home/opam/.opam/4.03/lib/hashcons/hashcons.cmx /home/opam/.opam/4.03/lib/result/result.cmxa /home/opam/.opam/4.03/lib/dns/dns.cmxa /home/opam/.opam/4.03/lib/lwt/lwt.cmxa /home/opam/.opam/4.03/lib/ocplib-endian/ocplib_endian.cmxa /home/opam/.opam/4.03/lib/ocplib-endian/bigstring.cmxa /home/opam/.opam/4.03/lib/mirage-profile/mirage_profile.cmxa /home/opam/.opam/4.03/lib/dns-lwt/dns_lwt.cmxa /home/opam/.opam/4.03/lib/ipaddr/ipaddr_unix.cmxa /home/opam/.opam/4.03/lib/lwt/lwt-log.cmxa /home/opam/.opam/4.03/lib/lwt/lwt-unix.cmxa lwt-unix/dns_lwt_unix.cmxa examples/.forward.eobjs/native/server.cmx)
# File "_none_", line 1:
# Error: Files lwt-unix/dns_lwt_unix.cmxa
#        and /home/opam/.opam/4.03/lib/dns-lwt/dns_lwt.cmxa
#        make inconsistent assumptions over implementation Dns_server
#     ocamlopt examples/forward.exe (exit 2)
# (cd _build/default && /home/opam/.opam/4.03/bin/ocamlopt.opt -w -40 -g -o examples/forward.exe -I /home/opam/.opam/4.03/lib/base64 -I /home/opam/.opam/4.03/lib/bytes -I /home/opam/.opam/4.03/lib/cstruct -I /home/opam/.opam/4.03/lib/dns -I /home/opam/.opam/4.03/lib/dns-lwt -I /home/opam/.opam/4.03/lib/hashcons -I /home/opam/.opam/4.03/lib/ipaddr -I /home/opam/.opam/4.03/lib/lwt -I /home/opam/.opam/4.03/lib/mirage-profile -I /home/opam/.opam/4.03/lib/ocplib-endian -I /home/opam/.opam/4.03/lib/re -I /home/opam/.opam/4.03/lib/re/posix -I /home/opam/.opam/4.03/lib/re/str -I /home/opam/.opam/4.03/lib/result -I /home/opam/.opam/4.03/lib/seq -I /home/opam/.opam/4.03/lib/sexplib -I /home/opam/.opam/4.03/lib/stringext -I /home/opam/.opam/4.03/lib/uri -I lwt-unix /home/opam/.opam/4.03/lib/ocaml/unix.cmxa /home/opam/.opam/4.03/lib/ocaml/bigarray.cmxa /home/opam/.opam/4.03/lib/sexplib/sexplib.cmxa /home/opam/.opam/4.03/lib/cstruct/cstruct.cmxa /home/opam/.opam/4.03/lib/seq/seq.cmxa /home/opam/.opam/4.03/lib/re/re.cmxa /home/opam/.opam/4.03/lib/re/str/re_str.cmxa /home/opam/.opam/4.03/lib/ipaddr/ipaddr.cmxa /home/opam/.opam/4.03/lib/re/posix/re_posix.cmxa /home/opam/.opam/4.03/lib/stringext/stringext.cmxa /home/opam/.opam/4.03/lib/uri/uri.cmxa /home/opam/.opam/4.03/lib/base64/base64.cmxa /home/opam/.opam/4.03/lib/uri/services.cmxa /home/opam/.opam/4.03/lib/hashcons/hashcons.cmx /home/opam/.opam/4.03/lib/result/result.cmxa /home/opam/.opam/4.03/lib/dns/dns.cmxa /home/opam/.opam/4.03/lib/lwt/lwt.cmxa /home/opam/.opam/4.03/lib/ocplib-endian/ocplib_endian.cmxa /home/opam/.opam/4.03/lib/ocplib-endian/bigstring.cmxa /home/opam/.opam/4.03/lib/mirage-profile/mirage_profile.cmxa /home/opam/.opam/4.03/lib/dns-lwt/dns_lwt.cmxa /home/opam/.opam/4.03/lib/ipaddr/ipaddr_unix.cmxa /home/opam/.opam/4.03/lib/lwt/lwt-log.cmxa /home/opam/.opam/4.03/lib/lwt/lwt-unix.cmxa lwt-unix/dns_lwt_unix.cmxa examples/.forward.eobjs/native/forward.cmx)
# File "_none_", line 1:
# Error: Files lwt-unix/dns_lwt_unix.cmxa
#        and /home/opam/.opam/4.03/lib/dns-lwt/dns_lwt.cmxa
#        make inconsistent assumptions over implementation Dns_resolver
```

File: packages/dns-async/dns-async.1.1.3/opam [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+and server resolver using the Jane Street Async library.
+"""
+license: "ISC"
+tags: ["org:mirage" "org:xapi-project"]
+homepage: "https://github.com/mirage/ocaml-dns"
+doc: "https://mirage.github.io/ocaml-dns/"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune"
```
Comment (kit-ty-kate) [inline packages/dns-async/dns-async.1.1.3/opam]:
```suggestion
  "dune" {>= "1.2"}
```

File: packages/dns-lwt-unix/dns-lwt-unix.1.1.3/opam [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+(as supported by Lwt_unix upstream).
+"""
+license: "ISC"
+tags: ["org:mirage" "org:xapi-project"]
+homepage: "https://github.com/mirage/ocaml-dns"
+doc: "https://mirage.github.io/ocaml-dns/"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune"
```
Comment (kit-ty-kate) [inline packages/dns-lwt-unix/dns-lwt-unix.1.1.3/opam]:
```suggestion
  "dune" {>= "1.2"}
```

File: packages/dns-lwt/dns-lwt.1.1.3/opam [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+opam package, and in theory you could compile this one to JavaScript
+as well using js_of_ocaml.
+"""
+tags: ["org:mirage" "org:xapi-project"]
+homepage: "https://github.com/mirage/ocaml-dns"
+doc: "https://mirage.github.io/ocaml-dns/"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+depends: [
+  "ocaml" {>= "4.03.0"}
+  "dune"
```
Comment (kit-ty-kate) [inline packages/dns-lwt/dns-lwt.1.1.3/opam]:
```suggestion
  "dune" {>= "1.2"}
```

File: packages/mirage-dns/mirage-dns.3.1.3/opam [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+homepage: "https://github.com/mirage/ocaml-dns"
+bug-reports: "https://github.com/mirage/ocaml-dns/issues"
+synopsis: "DNS implementation for the MirageOS unikernel framework"
+description: """
+This is an implementation of a DNS server and client resolver
+for the [MirageOS unikernel framework](https://mirage.io).
+"""
+depends: [
+  "ocaml" {>= "4.05.0"}
+  "dune"
```
Comment (kit-ty-kate) [inline packages/mirage-dns/mirage-dns.3.1.3/opam]:
```suggestion
  "dune" {>= "1.2"}
```

**Comment by avsm** [comment]
Thanks for the fixes @kit-ty-kate! Test results are fine -- I'll fix those separately later

File: packages/mirage-dns/mirage-dns.3.1.2/opam [hunk 10]
```diff
@@ -20,7 +20,7 @@ for the [MirageOS unikernel framework](https://mirage.io).
 depends: [
   "ocaml" {>= "4.05.0"}
   "dune"
-  "dns-lwt"
+  "dns-lwt" {>="1.1.2"}
```
Comment (samoht) [inline packages/mirage-dns/mirage-dns.3.1.2/opam]: What's the rational for the changes? this seems to cause a large breakage in the Mirage ecosystem as a lot of package cannot be co-installed anymore.

Comment (avsm) [reply packages/mirage-dns/mirage-dns.3.1.2/opam]: twas a mistake...

**Pull request merged**